### ISSUE TITLE
Considerations for locked states

### ISSUE THREAD

**Issue by awrichar** [hyperledger-labs/zeto#119]
In reviewing the latest "lock states" functionality (added in #110), I'm trying to understand the current overall state model of Zeto, and whether it can be considered a UTXO model or something else entirely.

For the most part, Zeto follows UTXO - states have a value and an owner, and they can be spent exactly once as an input to a transaction. However, with the addition of "locked states", each state can now have additional metadata specifying whether or not the state is "locked". In the UTXO world, this seems similar to a "spending rule" or "script", as used in systems like Bitcoin and Cardano. However, the Zeto implementation appears to be novel because you can attach or alter this metadata _without spending the state_. Consider:

* `A` - $100 coin, spendable by whoever can generate a valid proof
* `A'` - same as A, but only spendable by one specific address

As of #110, it's possible to transform `A` to `A'` without spending `A`. And from there it may transform back to `A`, or to `A''`, etc. This presents problems when trying to index Zeto coins. A normal UTXO indexer deals in very "final" operations - a state comes into existence exactly once, and gets spent exactly once. With this notion of mutable spending rules, a much more sophisticated indexer would be needed - one which tracks that `A` may become spendable or not spendable at various moments throughout its lifetime, based on a set of changing rules.

Given this, I believe there is a fundamental choice to be made in terms of Zeto's future design around "locks":
1. We conform to UTXO, meaning that `A` is spent to generate `A'`, and we explore the implications for ZKPs under this constraint (for instance, can we create a proof that is valid to spend either `A` or `A'`, because it relies only on the owner/value of `A` and not on the "locked" status?)
2. We design a novel system of "UTXO with mutable spending rules", and figure out the implications for building a robust indexer for such a system

**Comment by awrichar** [comment]
@jimthematrix I don't think nullifiers change the execution graph of the UTXO model - they just change the mechanism by which you spend. A state is only spendable if you can generate the proper nullifier for it. To facilitate this, in Paladin we have taught it about this specific spending rule, and it is turned on at a contract level (to say "all states in this contract must be spent by generating nullifiers"). Another option of course would be to have more "generalized" spending rules spelled in a DSL, such as other UTXO models have adopted - but that's a bit lift and we haven't yet decided to take it on.

But that's a huge difference from states that change their spending rules on the fly. I can't have a state that was generated without requiring nullifiers, and _change the spending rule_ to say that from this point, it will have to be spent with a nullifier instead. This is the big difference I see with the current locking behavior - the rules around spending UTXOs are not fixed throughout the life of the UTXO, but they can change independent of the UTXO itself.

**Comment by jimthematrix** [comment]
Correct, which _I think_ is the same when a UTXO is spent and the new UTXO in the output has different spending rules than the ones that just got spent. The indexer still has to react on a per UTXO basis based on each UTXO's spending rules. Not sure I see yet a big difference b/w the current implementation of keeping the same UTXO id but updating the spending rules vs. the proposal to spend and create a new UTXO with a different spending rule and a different ID.

**Comment by awrichar** [comment]
I think the following are generally needed to be able to index UTXOs:
1. Each UTXO can be uniquely identified
2. Each UTXO can only be spent once, and can't go back to "unspent" after being spent

I believe you're saying that to satisfy (1), an indexing system could elect to identify UTXOs by the union of (UTXO id, locked status). Therefore if `A` transitions to `A'` by changing the locked status but not the on-chain id of the UTXO, the indexer could consider that `A` is spent and `A'` is a new state.

However, if `A'` is altered to remove the "locked" status, this would turn it back into `A` again, which would involve `A` transitioning from "spent" to "unspent", and would violate (2) above.

I think (2) is a pretty fundamental rule of a UTXO system, so I'm having trouble coming up with an indexing model that doesn't require it to be true.

**Comment by jimthematrix** [comment]
No argument with the above assertions. I think we are narrowly defining all state transitions in an UTXO systems as spending, so there can be no other types of state transitions such as locking which can be reverted without spending the UTXO.

I tried to search for formal UTXO definitions in the academia space, such as researchgate.net and mdpi.com, but haven't found any unfortunately. If there are such definitions that clearly spells out the above assumptions, I think we have an easy decision to make.

In the absence of that, I'm trying to discover the benefit of the above assumption. One obvious benefit is the fact that it leads to a DAG type state transition graph that is more efficient to manage than the graphs that have cycles. For instance in the DB storage design for implementing a Zeto client, we can have an append-only scheme, rather than relying on updates. Such schemes result in higher performance. This is the current design of [Paladin](https://github.com/LF-Decentralized-Trust-labs/paladin).

**Comment by awrichar** [comment]
It's not particularly academic, but I asked ChatGPT about it and generally got what I'd expect: https://chatgpt.com/share/6776f6c1-c090-8012-9338-ce1d73120415

**Comment by jimthematrix** [comment]
```
In the UTXO model, scripts cannot be changed on the fly. Instead, to modify the conditions or script associated with a UTXO, you must:

- Spend the UTXO (fulfilling its current script requirements).
- Create a new UTXO with a different script.

This is because UTXOs are immutable once created. The immutability ensures the integrity and security of the blockchain, as any changes to a UTXO would undermine the consensus rules.
```

and

```
Why UTXOs Can’t Change Scripts on the Fly
- Immutability: UTXOs are a fundamental part of the blockchain's ledger state. Changing them directly would break the chain of cryptographic proofs linking transactions.
- Determinism: The UTXO model relies on deterministic validation. Every UTXO’s state must be reproducible across all nodes in the network. Allowing mutable scripts would introduce complexity and potential inconsistencies.
- Simpler Validation: Keeping UTXOs immutable ensures that nodes can easily verify whether a UTXO is valid without tracking dynamic state changes.
```

I think the above are the most relevant part of the ChatGPT response. However they are in the context of a blockchain protocol based on UTXO. In a UTXO system built on EVM, immutability and determinism are guaranteed by the underlying protocol (account based) already. So I don't think having a UTXO design that can alter its spending rules fundamentally breaks anything (again this is without having authoritative sources on what UTXOs should be).

**Comment by jimthematrix** [comment]
That said, I'm leaning toward changing the lock mechanism to spending rather than modifying as you have proposed, for the benefits discussed so far related to client implementations.

**Comment by awrichar** [comment]
Right - we definitely could have a system that is UTXO-like, and yet has more complex rules (such as having states other than spent/unspent, and more allowable transitions between states). Would it still be called UTXO? Maybe. Like you said, there's really no authority that will determine yes or no. That's a more academic question when you get down to it.

I'm mostly concerned with 1) properly describing the state model to others, and 2) making it easy for anyone to efficiently index Zeto contracts. Adhering to basic UTXO and ensuring indexing can be built in an append-only fashion definitely help on these counts.

**Comment by Chengxuan** [comment]
This is an interesting discussion. 
 
Here is what's Zeto today as per my understanding:

- (**base**). Zeto follows UTXO model
- (**locking**). Zeto provides an extra function (optional) to record locking states on chain. 
    - Zeto has other optional functions (e.g. kyc, erc20 anchoring)

Also, locking can be implemented without using this optional function:

- (**escrowlocking**). transfer the amount to an escrow address  (which also record the locking states on chain)
- (**offchainlocking**). UTXO owner have an off-chain web3 app to manage the locking states. (locking states are not recorded on chain)

In the conversation, I saw the following threads that have been discussed:
1. exploring whether **locking** should become a required feature (part of **base**) by influencing the definition of `UTXO` model
2. have better/more formal documentation on the interaction pattern when the **locking** is used
3. remove **locking**, and encourage **escrowlocking** to be the best practice.


I think `3` is probably OK for designing a coin that follows the existing UTXO model. `2` is a stepping stone towards `1` and is an interesting area to explore, however, I don't think it's a Zeto-specific thing (at least at the interface level).

**Comment by awrichar** [comment]
@Chengxuan I think some form of locking and delegation _must_ be standardized and provided in base Zeto. Otherwise there will be no standard way to implement any form of atomic settlement using Zeto.

Zeto tokens can only be held by EOAs, as you must generate BJJ keys and ZKPs in order to move them. They can't be held by other smart contracts in the way that ERC20 can, for example. Therefore, the only way to provide programmable atomic constructs is to let an EOA set up the proofs and states, and lock them to be finalized by a particular address (which may be an EOA or smart contract).

**Comment by awrichar** [comment]
@Chengxuan correct. But with Zeto I think there's no way to say "A can spend any amount of my tokens in any number of different transactions, totaling up to a max of 80" - at least not in a way that is guaranteed in the smart contract. @jimthematrix is more of an authority on what's possible with the ZKPs, but because of the way value is tracked, I don't think we can do this in the same way it's done on ERC-20.

What we can do is say "I approve A to spend exactly this set of UTXO states", and pre-select a set of states totaling 80. We've landed on "locking" the states while the approval is in effect, simply because it's much too easy to accidentally go and spend those states on something else and therefore invalidate the approval (that was the focus of #110). You could argue for a less reliable form of approval where A has no idea if the states will continue to be valid or not - I'm not saying it's useless, but I don't know under what scenario you would want that instead.

There's another form of delegation sometimes called "approval for all" that's used in ERC-1155 and some other standards. It's something we haven't explored in detail, but I think Zeto could potentially support this pattern of delegation as well, if there was a use for it. But it would be a different ZKP.

In summary I think we have:
1. approve someone to spend a specific set of states
2. approve someone to spend any arbitrary set of states totaling up to a maximum
3. approve someone to spend any of my states

I believe 1 & 3 are possible, and 2 is not possible.

**Comment by awrichar** [comment]
Yes, I believe this aligns with where we're going. You've summarized mostly how the Noto implementation of locks is now proposed to work on Paladin, and based on discussion here and https://github.com/hyperledger-labs/zeto/issues/118, it feels like Zeto is trending in a similar direction ("locking" means spending the inputs and generating some number of "locked" outputs along with a refunded "remainder", if necessary). All preparation of input/output states and proofs has to be done by a client, but spending of locked value can be delegated/finalized by a smart contract address.

One major goal here is to enable Zeto to be atomically swapped with any other token (either an ERC20, or Noto, etc). So you're exactly right with `enabling a smart contract address to spend Zetos`, and additionally we want to be sure Zeto offers a way to give some counterparty protection in the window between when I delegate spending ability, and when the spending happens.

**Comment by jimthematrix** [comment]
agreeing with the above conversation @Chengxuan @awrichar. No additional comments to add.

Something I'd like to call out as a potential significant decision as I got further with the idea to "spend the UTXOs and create locked UTXOs in the output". @awrichar and I have had some preliminary discussions about this.

When I initially implemented the new locking function for the non-nullifier token implementations, both for fungible and non-fungible tokens, it was pretty straightforward to do:
- the locked UTXOs are managed in the same map as regular (unlocked) UTXOs, but additionally marked as locked in a separate map for all locked UTXOs
- the lock() function is a regular transfer plus adding the locked outputs to the locked UTXOs map
- the transfer() function is enhanced to check that the inputs are either unlocked, or are locked and the transaction submitter is the designated delegate according to the locked UTXOs map

This works pretty well, with a small overhead in all transfer() calls, to check the inputs against the locked UTXOs map. The benefit is that we can use a single transfer() function for consuming both unlocked and locked states.

However when I got to tokens using nullifiers, I realized that to keep using a single transfer() function, it would have to proof two things:
1. that the nullifiers are bound to UTXOs in the regular UTXOs merkle tree (which is already included in the proof statements), aka inclusion proof
2. that the nullifiers are bound to UTXOs NOT in the locked UTXOs merkle tree, aka non-inclusion proof

Both statements are required for every transfer() call. This is going to make the transfer() function very expensive, especially on the prover side.

On the other hand, if we maintain the locked UTXOs in a separate list, and have two separate functions for spending unlocked vs. locked UTXOs, then each transfer function only needs to deal with one inclusion proof.

I'm leaning toward tracking unlocked and locked UTXOs separately and having two separate functions for spending unlocked vs. locked UTXOs.
